### PR TITLE
Upload script generator task

### PR DESCRIPTION
**Pull request by joelbyler**
This is a rough spike but I wanted to go ahead and create the PR in the hopes to start the conversation.  I borrowed some of this from generators for phoenix.  Please feel free to comment on anything here I plan to do some refactoring to the task and extract a few functions so that its not so procedural.  

One question that's on my mind presently is that I'm wondering how much of the file can be interpolated using prompts when generating the file instead of relying on environment variables.  I'm using eex templates so we should be able to prompt for values and interpolate them into the script if that's desirable.  

Secondary question, I verbatim copied `upload.sh` into the `priv/templates` obviously I don't think having this duplication will work long run, if this gets merged in, will there be any advantage to leaving the original upload script in place at the project root or should we drop it?

Thanks, looking forward to feedback!

### CODE REVIEW

**Comment by fhunleth** [comment]
Cool. Thanks!

Regarding the first question on replacing the environment variables, I'm leaning on keeping the environment variables since I know of a few projects where people switch between boards and triggering off environment variables is convenient. 

Regarding the second question, when this gets merged in, I'm going to delete the `upload.sh` in the root directory. Or feel free to update the PR to delete it. If I don't do this, I'll certainly have one get out of date.

One thing that would be good is to update the `README.md` with the new way of downloading `upload.sh`. Could you add something short?

**Comment by joelbyler** [comment]
@fhunleth thanks for the feedback.  I've updated the PR and squashed into a single commit.  Let me know if there's anything else needed for this.  If there are values that you think would be good to prompt for I can add those but per your earlier feedback I've mostly left the script in tact except that its now an eex template, hopefully thats ok.

File: lib/mix/tasks/nerves.gen.script.upload.ex [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      filename = "upload.sh"
+    end
+    upload_script_contents = EEx.eval_file("./priv/templates/script.upload.eex", [])
+    Mix.shell.info """
+
+    Generating a new file named #{filename}
+    """
+    File.write(filename, "#{upload_script_contents}\n")
+
+    System.cmd("chmod", ["+x", filename])
```
Comment (fhunleth) [inline lib/mix/tasks/nerves.gen.script.upload.ex]: Could you change this to use `File.chmod!(filename, 0o755)`?

**Comment by fhunleth** [comment]
Looks good. Assuming no issues with using `File.chmod!/2`, I'll merge after class today. Thanks!

File: lib/mix/tasks/nerves.gen.script.upload.ex (hunk 1 above)
Comment (joelbyler) [reply lib/mix/tasks/nerves.gen.script.upload.ex]: Yep, I can do that. I'm also seeing a warning when I run so I'll try and clean that up as well.

**Comment by joelbyler** [comment]
@fhunleth I've updated per your comments.  Also, I was going to write some tests for this but having trouble actually running `mix test`, and seeing an error: `failed to start child: Nerves.Runtime.Kernel.UEvent`  how are you running these?  It must be magic ✨ !  😄    Perhaps an update to the readme for developers who want to contribute would be helpful.  I can add that to another PR if you can give me some advice or maybe the test is overkill? Thanks!

**Comment by fhunleth** [comment]
Merged!

Also, I added a bug regarding the unit testing issue. See https://github.com/nerves-project/nerves_runtime/issues/34. A unit test for the generator would be nice, but I didn't want to hold up merging this feature especially with the OSX uevent issue. If you have easy access to Linux, then you won't see the problem, but no worries if not.

**Pull request merged**